crates/dht/src/time.rs

```diff
@@ -0,0 +1,1034 @@
+                        + Duration::from_secs(
+                            i * (1u64 << self.factor) * UNIT_TIME.as_secs(),
+                        ),
+                )
+                .await?;
+
+            let hash = combine_op_hashes(op_hashes);
+
+            store
+                .store_slice_hash(self.full_slices, hash.to_vec())
+                .await?;
+
+            self.full_slices += 1;
+        }
+
+        full_slices_end_timestamp += Duration::from_secs(
+            new_full_slices_count * (1u64 << self.factor) * UNIT_TIME.as_secs(),
+        );
+
+        let new_partials =
+            self.layout_partials(current_time, full_slices_end_timestamp)?;
+        let old_partials = std::mem::take(&mut self.partial_slices);
+
+        for (start, size) in new_partials.into_iter() {
+            // If this slice didn't change, then we can reuse the hash
+            let maybe_old_hash = old_partials.iter().find_map(|b| {
+                if b.start == start && b.size == size {
+                    Some(b.hash.clone())
+                } else {
+                    None
+                }
+            });
+
+            // Otherwise, we need to get the op hashes for this time slice and combine them
+            let hash = match maybe_old_hash {
+                Some(h) => h,
+                None => {
+                    let end = start
+                        + Duration::from_secs(
+                            (1u64 << size) * UNIT_TIME.as_secs(),
+                        );
+                    combine_op_hashes(
+                        store
+                            .retrieve_op_hashes_in_time_slice(start, end)
+                            .await?,
+                    )
+                }
+            };
+
+            self.partial_slices.push(PartialSlice { start, size, hash })
+        }
+
+        // There will be a small amount of time left over, which we can't partition into smaller
+        // slices. Some amount less than [UNIT_TIME] will be left over.
+        // Once that has elapsed up to the next [UNIT_TIME] boundary, we can update again.
+        if let Some(last_partial) = self.partial_slices.last() {
+            self.next_update_at = last_partial.end() + UNIT_TIME;
+        }
+
+        Ok(())
+    }
+}
+
+// Private methods
+impl PartitionedTime {
+    /// Private constructor, see [PartitionedTime::from_store].
+    ///
+    /// This constructor just creates an instance with initial values, but it doesn't update the
+    /// state with full and partial slices for the current time.
+    fn new(origin_timestamp: Timestamp, factor: u8) -> Self {
+        Self {
+            origin_timestamp,
+            factor,
+            full_slices: 0,
+            partial_slices: Vec::new(),
+            // This only changes based on the factor, which can't change at runtime,
+            // so compute it on construction
+            min_recent_time: residual_duration_for_factor(factor - 1),
```

Comment:
why -1 here?

Reply:
I can't change this actually, because it's also used literally in the calculation code, so I'd end up adding an equally strange `+ 1`. So I'll add some information into the code comment about why that's being done

Reply:
Hehe, cool, that works for me.